language/tools/move-mv-llvm-compiler/src/stackless/llvm.rs

```diff
@@ -761,15 +778,21 @@ impl Builder {
         }
     }
 
-    // TODO: If \p name isn't provided get a tempname.
     pub fn build_binop(
         &self,
         op: LLVMOpcode,
         lhs: AnyValue,
         rhs: AnyValue,
         name: &str,
     ) -> AnyValue {
-        unsafe { AnyValue(LLVMBuildBinOp(self.0, op, lhs.0, rhs.0, name.cstr())) }
+        assert!(name.trim().len() == name.len(), "Invalid name provided for identifier");
+        let tempop = &LLVMOpcode1(op).to_string().to_ascii_lowercase()[4..]; // Strip LLVM prefix.
+        let opname = if name.is_empty() {
+            tempop
+        } else {
+            name
+        };
+        unsafe { AnyValue(LLVMBuildBinOp(self.0, op, lhs.0, rhs.0, opname.cstr())) }
     }
```

Comment:
This seems to be a draft patch or work in progress, but I'll add comments since it was requested. 

This has some problems. The struct `LLVMOpcode1` and associated functions are completely unnecessary and should be removed. But even if those were kept, the code above printing it would result in strange output. That is, not just the llvm opcode, but something like `opcode1(llvmfoo)` which is surely not what was wanted. It should also use `strip_prefix` instead of indexing the string.

This entire patch can be replaced by just a few lines (that, along with a `move-ir-tests` testcase):
```
     ) -> AnyValue {
-        unsafe { AnyValue(LLVMBuildBinOp(self.0, op, lhs.0, rhs.0, name.cstr())) }
+        let opname = if name.is_empty() {
+            format!("{:?}", op).strip_prefix("LLVM").unwrap().to_lowercase().cstr()
+        } else {
+            name.cstr()
+        };
+        unsafe { AnyValue(LLVMBuildBinOp(self.0, op, lhs.0, rhs.0, opname)) }

```

Another approach, which is overkill for something so trivial, is to use LLVM's `Instruction::getOpcodeName`, but it is not currently exposed by `llvm_sys`. We could add it to `llvm-extra-sys` and call it from `binop`. The only possible benefit to that overkill is that the output names are stable and directly from LLVM itself.

Reply:
Finally, note that it is rare that `build_binop` is called with an empty name. In the entire translator, there are only two instances. One in `translate_vector_comparison_impl` and the other in `translate_struct_comparison_impl` where we invert the sense of the comparison:
```
dst_reg = self.module_cx.llvm_builder.build_binop(
                llvm_sys::LLVMOpcode::LLVMXor,
                dst_reg,
                cval,
                "",
            );
```
It would be trivial to simply provide a name here (like "invert")-- no other patching necessary.

Reply:
> i guess it is better to provide names in those instances, and add assert in build_binop then?

- Yes, it is better to simply provide the names in the two functions mentioned, and drop the patch altogether.
- If the current patch is retained at all, I would write it as I have shown above-- that would be the whole patch, that is literally all that is "needed."
- Also, there also is no need for an assert in `build_binop`. The lack of a name is not an error. See the way I have rewritten things above. As mentioned, the lack of names is rare and inconsequential.
- Note that `Instruction::getOpcodeName` is also not exposed in the LLVM-C API. I believe `llvm_sys` only wraps LLVM-C APIs, so it would first also need a patch in LLVM proper. And we'd have to update our local LLVM once all of that is done. This is why I mentioned adding to `extras`. But again, this is totally unnecessary, total overkill.